Use external data for home page content cards

pmac opened the pull request:
This is a WIP for the home page external admin work. @alexgibson let me know what you think of this first pass when you get some time.

Re #6124

pmac commented:
I still have to deal with image uploading and linking, but this is a decent start I think.

pmac commented:
It's working for me locally if I do the following:

* `make run`
* open a new terminal also to bedrock repo
* `docker-compose exec app bash`
* `./manage.py migrate`
* `./manage.py update_content_cards`
* then load the dev server in a browser

The image will be broken, but everything else about card 1 should work.

alexgibson commented:
I ran the steps above and everything worked as expected 👍 

Some thoughts on GA and l10n:

- I'm not sure we need the `ga_title` field in the external repo or in the DB. The field itself is just to store an English version of the card title for GA purposes, however in the world of this type of automation, I think it's pretty safe that this will all be in English anyway, and we can just use `title` to populate the data attribute in the template macro.
- If we do go on to support l10n in this system, I think it wil be unlikely that we would support string translations in the typical sense like we do on bedrock. That kind of thing will be for evergreen content, and not something in www-admin. 
- What I do think we need to support however, is for specific locales to have their own content. For example we do want to build a German version of the home page, which will have its own curated content and links to German blog posts (See https://github.com/mozilla/bedrock/issues/6585).

What we have here so far looks great, let me know if you would like to chat this week about the above. Thanks @pmac!

pmac commented:
Great! I've basically made it open ended. I think it's just a convenience to have all locales of a specific card inherit the attributes from the en-US one so that it's easier to update them all if that's what you need to do (change the image e.g.), but if that feels over-engineered and potentially error prone I'm happy to remove it. The way it basically works now is that you can provide data for any argument to the `card()` macro and it'll pass through. Some are handled in a special way (e.g. just saying `aspect_ratio: 16-9` and have that become the right class name) but basically that's it.

And yes, I'd love to have a chat this week to further sync about the status and future. Thanks for the review!

pmac commented:
Changed recently:

* Added tests for importing the card data and the model functions.
* Improved how the template functions work to allow it to get the cards from the context or to be passed in.
* Fixed ability to provide either a full URL or URL name in the data

pmac commented on bedrock/base/templates/macros-protocol.html:
```diff
@@ -95,11 +95,15 @@
   <a class="mzp-c-card-block-link" href="{{ link_url }}" data-link-name="{{ ga_title }}" data-link-type="link" data-link-group="card">
     <div class="mzp-c-card-media-wrapper">
-      {{ lazy_img(image_url=image_url, include_highres_image=include_highres_image, placeholder_url='placeholder.png', optional_attributes={'class': 'mzp-c-card-image'}) }}
+      {% if highres_image_url %}
```
Comment added. LMK if what I wrote is clear and helpful enough. Good suggestion 👍

alexgibson commented on .env-dist:
```diff
@@ -2,3 +2,5 @@ DEBUG=True
 ALLOWED_HOSTS=*
 SWITCH_NEWSLETTER_MAINTENANCE_MODE=False
+CSP_DEFAULT_SRC='self',*.mozilla.net,*.mozilla.org,*.mozilla.com,*.s3.amazonaws.com
+CONTENT_CARDS_URL=http://bedrock-dev-media.s3.amazonaws.com/media/
```
Do we need to add this variable for demo deployments?

pmac replied:
oooooooooo... good point. I think we should point demos to https://www-dev.allizom.org/media/. Should probably use that both places... I'll do that. Good catch.

alexgibson commented on .env-dist:
```diff
@@ -2,3 +2,5 @@ DEBUG=True
 DEV=True
 ALLOWED_HOSTS=*
 SWITCH_NEWSLETTER_MAINTENANCE_MODE=False
+CSP_DEFAULT_SRC='self',*.mozilla.net,*.mozilla.org,*.mozilla.com,*.allizom.org
```
When I run the site locally (not using Docker), this line seems to prevent all assets from loading? I also see things like `Content Security Policy: Couldn't parse invalid host self'` in the web console.

Things seems to be fine when using Docker, which is odd.

pmac replied:
@alexgibson I believe I've fixed this. It was an issue with the quotes around `'self'` and getting that out of the environment. Our library to do that strips quotes. So I moved to only neededing to provide the "extra" items to add to the `CSP_DEFAULT_SRC` list. This seems better anyway.

alexgibson commented on docs/content-cards.rst:
```diff
@@ -0,0 +1,69 @@
+
+.. note::
+
+    The ``get_page_content_cards`` function is not all that clever as far as l10n is concerned. If you have
```
> you should have cards for every locale for which the page is active or the function will return an empty dict.

I'm not sure if having cards translated for every locale a page is active for is something we can realistically try to adhere to? Locales can be activated at any time, independent of changes to either bedrock or the www-admin repo. The warning you have here is great, but should we also be strongly encouraging developers to be explicit when passing `locale` into `get_page_content_cards`?

pmac replied:
It's a tough question/problem. Are we prepared to say that a template using these should only be locale specific, and then hardcode? That's fine, but a bit limiting. Perhaps we should recommend wrapping the cards section of the template in a language specification at least? 

```jinja
<% if LANG in ['de', 'fr', pt-BR'] %>
{# do cards #}
<% endif %>
```

Or decide in the view and only do cards if the `page_content_cards` variable is truthy?

pmac replied:
I updated the docs with an attempt. LMK what you think.

alexgibson replied:
Works for me 👍

pmac commented:
@alexgibson I believe I've addressed the things you've found. I've also rebased on current master and pushed my demo again. I also just updated the data in the www-admin repo with the new cards on the current home page. I need to squarsh, but other than that we might be good to go here.

pmac commented:
demo is up https://bedrock-demo-pmac.oregon-b.moz.works/en-US/

alexgibson commented on .env-dist:
```diff
@@ -2,3 +2,5 @@ DEBUG=True
 DEV=True
 ALLOWED_HOSTS=*
 SWITCH_NEWSLETTER_MAINTENANCE_MODE=False
+CSP_DEFAULT_SRC=*.allizom.org
```
Hmm, when running locally (not using Docker) I still see CSP issues e.g.

```
Content Security Policy: The page's settings blocked the loading of a resource at http://localhost:8000/media/js/BUNDLES/site.js ("script-src").
```

pmac replied:
@alexgibson it's working for me. I honestly have no idea. Does the `CSP_DEFAULT_SRC` setting in your `.env` file look exactly like this line?

alexgibson replied:
I tried this again with a fresh `.env` and it worked both locally and using docker. Not sure what I did wrong the first time, but I can't seem to replicate this error any longer? :)
